android/src/main/java/com/urbanairship/reactnative/EventEmitter.java

```diff
@@ -46,110 +45,110 @@ static EventEmitter shared() {
-            return;
-        }
-
-        ReactContext reactContext = reactInstanceManager.getCurrentReactContext();
-        if (reactContext != null && reactContext.hasActiveCatalystInstance()) {
-            emit(reactContext, event.getName(), event.getBody());
-        } else if (reactInstanceManager.hasStartedCreatingInitialContext()) {
-            reactInstanceManager.addReactInstanceEventListener(new ReactInstanceManager.ReactInstanceEventListener() {
-                public void onReactContextInitialized(ReactContext reactContext) {
-                    emit(reactContext, event.getName(), event.getBody());
-                    reactInstanceManager.removeReactInstanceEventListener(this);
-                }
-            });
-        }
+            }
+        });
     }
 
     /**
      * Called when a new listener is added for a specified event name.
      *
      * @param eventName The event name.
      */
-    void addAndroidListener(ReactContext reactContext, String eventName) {
+    void addAndroidListener(String eventName) {
         synchronized (knownListeners) {
-            for (Event event : new ArrayList<>(pendingEvents)) {
-                if (event.getName().equals(eventName)) {
-                    sendEvent(reactContext, event);
-                    pendingEvents.remove(event);
-                }
-            }
             listenerCount++;
             knownListeners.add(eventName);
         }
+
+        mainHandler.post(new Runnable() {
+            @Override
+            public void run() {
+                sendPendingEvents();
+            }
+        });
     }
 
     /**
      * Called when listeners are removed.
      *
      * @param count The count of listeners.
      */
-    void removeAndroidListeners(ReactContext reactContext, int count) {
+    void removeAndroidListeners(int count) {
         synchronized (knownListeners) {
-            long currentCount = listenerCount;
-            listenerCount = max(0, currentCount - count);
-
-            if (listenerCount == 0) {
+            listenerCount -= count;
+            if (listenerCount <= 0) {
+                listenerCount = 0;
                 knownListeners.clear();
             }
         }
     }
 
     /**
-     * Helper method to emit data.
-     *
-     * @param reactContext The react context.
-     * @param eventName The event name.
-     * @param eventBody The event body.
+     * Attempts to send pending events.
      */
-    private void emit(ReactContext reactContext, String eventName, Object eventBody) {
-        reactContext.getJSModule(RCTNativeAppEventEmitter.class).emit(eventName, eventBody);
+    @MainThread
+    private void sendPendingEvents() {
+        synchronized (knownListeners) {
+            for (Event event : new ArrayList<>(pendingEvents)) {
+                if (knownListeners.contains(event.getName())) {
+                    pendingEvents.remove(event);
```

Comment:
I'm a little confused by this. We're removing the the pending event here, and then in sendEvent we're adding it back if there are no valid listeners or emitting failed? Did I get that right?

Reply:
yeah, ill add a comment to make that clear